nixos/lib/test-driver/test_driver/driver.py

```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+                del self.machines[idx]
+
+            # add and change new machines
+            for start_cmd in start_cmds:
+                existing = [
+                    m for m in self.machines if m.name == start_cmd.machine_name
+                ]
+                if len(existing) == 0:
+                    machine = self.create_machine(start_cmd._cmd)
+                    name = pythonize_name(machine.name)
+                    globals()[name] = machine
+                    self.machines.append(machine)
+
+                    self.logger.info(
+                        f'new machine created {start_cmd.machine_name}. start it with driver.machines["{start_cmd.machine_name}"].start()'
+                    )
+                elif len(existing) == 1:
+                    machine = existing[0]
+                    if machine.start_command._cmd == start_cmd._cmd:
+                        self.logger.info(f"{start_cmd.machine_name} unchanged.")
+                    elif not machine.booted:
+                        self.logger.info(f"{start_cmd.machine_name} changed.")
+                        machine.start_command = start_cmd
+                    else:
+                        with self.logger.nested(
+                            f"{start_cmd.machine_name} updated. switching to new configuration"
+                        ):
+                            # TODO: assuming things are at well-known paths isn't great
+                            switch_cmd = (
+                                Path(start_cmd._cmd).parent.parent
+                                / "system"
+                                / "bin"
+                                / "switch-to-configuration"
+                            )
+                            machine.succeed(f"{switch_cmd} test")
+                            machine.start_command = start_cmd
+                else:
+                    self.logger.warning(
+                        f"{start_cmd.machine_name} has multiple instances. This shouldn't be possible, but either way it's now ambiguous which to update."
+                    )
+
+        with self.logger.nested("updating vlans"):
+            nrs = list(set([int(vlan) for vlan in vlans_str.split(" ")]))
+            to_del = []
+            for idx, vlan in enumerate(self.vlans):
+                if vlan.nr not in nrs:
+                    to_del.append(idx)
+            for idx in sorted(to_del, reverse=True):
+                self.logger.info(
+                    f"vlan {self.vlans[idx].nr} removed from test. Deleting it"
+                )
+                del self.vlans[idx]
+            # add new vlans
+            old_nrs = [vlan.nr for vlan in self.vlans]
+            for nr in nrs:
+                if nr not in old_nrs:
+                    self.vlans.append(VLan(nr, tmp_dir, self.logger))
+
+        # if the test script has already been run we can't post-hoc modify the actions, but we can at least update the script
```

Comment:
Maybe log this?

Reply:
Does this mean `test_script()` in the repl will launch the new script?

Reply:
Yep, that's the idea.

Reply:
Updated the log message and removed the comment.